src/lib/homestore_backend/hs_blob_manager.cpp

```diff
@@ -1,17 +1,220 @@
-    return folly::makeUnexpected(BlobError::UNKNOWN);
+void HSHomeObject::on_blob_put_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                      const homestore::MultiBlkId& pbas,
+                                      cintrusive< homestore::repl_req_ctx >& hs_ctx) {
+    repl_result_ctx< BlobManager::Result< BlobInfo > >* ctx{nullptr};
+    if (hs_ctx != nullptr) {
+        ctx = boost::static_pointer_cast< repl_result_ctx< BlobManager::Result< BlobInfo > > >(hs_ctx).get();
+    }
+
+    auto msg_header = r_cast< const ReplicationMessageHeader* >(header.bytes);
+
+    std::shared_lock lg{_pg_lock};
+    HS_PG* pg{nullptr};
+    if (auto it = _pg_map.find(msg_header->pg_id); it != _pg_map.end()) {
+        pg = s_cast< HS_PG* >(it->second.get());
+    } else {
+        LOGERROR("Couldnt find pg {} for blob {}", msg_header->pg_id, lsn);
+        ctx->promise_.setValue(folly::makeUnexpected(BlobError::UNKNOWN_PG));
+        return;
+    }
+
+    BlobInfo blob_info;
+    blob_info.shard_id = msg_header->shard_id;
+    blob_info.blob_id = lsn;
+    blob_info.pbas = pbas;
+
+    // Write to index table with key {shard id, blob id } and value {pba}.
+    auto r = add_to_index_table(pg, blob_info);
+    if (r.hasError()) {
+        LOGERROR("Failed to insert into index table for blob {} err {}", lsn, r.error());
+        ctx->promise_.setValue(folly::makeUnexpected(r.error()));
+        return;
+    }
+
+    if (ctx) { ctx->promise_.setValue(BlobManager::Result< BlobInfo >(blob_info)); }
+}
+
+BlobManager::Result< Blob > HSHomeObject::_get_blob(ShardInfo const& shard, blob_id_t blob_id) const {
+    auto pg_id = shard.placement_group;
+    HS_PG* pg{nullptr};
+    {
+        std::shared_lock lg{_pg_lock};
+        if (auto it = _pg_map.find(pg_id); it != _pg_map.end()) {
+            pg = s_cast< HS_PG* >(it->second.get());
+        } else {
+            LOGWARN("failed to put blob with non-exist pg [{}]", pg_id);
+            return folly::makeUnexpected(BlobError::UNKNOWN_PG);
+        }
+    }
+
+    if (!pg->repl_dev_) {
+        LOGWARN("failed to get repl dev instance for pg [{}]", pg_id);
+        return folly::makeUnexpected(BlobError::PG_NOT_READY);
+    }
+
+    auto r = get_from_index_table(pg, shard.id, blob_id);
+    if (!r) {
+        LOGWARN("Blob not found in index id {} shard {}", blob_id, shard.id);
+        return folly::makeUnexpected(r.error());
+    }
+
+    auto multi_blkids = r.value();
+    auto block_size = pg->repl_dev_->get_blk_size();
+    auto sgs_ptr = std::make_shared< sisl::sg_list >();
+    auto total_size = multi_blkids.blk_count() * block_size;
+    auto iov_base = iomanager.iobuf_alloc(block_size, total_size);
+    sgs_ptr->size = total_size;
+    sgs_ptr->iovs.emplace_back(iovec{.iov_base = iov_base, .iov_len = total_size});
+
+    auto result = pg->repl_dev_->async_read(multi_blkids, *sgs_ptr, total_size).get();
```

Comment:
Again, defer everything that follows this.

Reply:
Ack. same as above